### ISSUE TITLE
Automatically identify any stream type

### ISSUE THREAD

**Issue by regatton** [google/ExoPlayer#3165]
I'm making a radio app for Android, and I'm looking into ExoPlayer as an alternative to the standard MediaPlayer because it has some limitations (eg. can't play AAC streams). Basically I have a list of radio URLs, and when the user selects a radio I feed the corresponding URL to ExoPlayer.

Unfortunately it doesn't seem to work on all streams... Some are `.pls` for which I guess I have to extract the mp3 URLs myself. Some AAC streams don't work (but they work on VLC or Firefox), and no HLS streams work. From the documentation, it looks like I need some special code for HLS and other formats. This is pretty tedious compared to simpler way of just feeding a URL to the player. Can ExoPlayer automatically detect the right format and use the corresponding extractor? Why do I have to detect the type of some special cases, when most format are already detected automatically?

**Comment by AquilesCanta** [comment]
> No HLS stream work

This is possibly because an `ExtractorMediaSource` is being used instead of an `HlsMediaSource`.

If you don't have control over the media to be played, it will be necessary to sniff at `MediaSource` level. Otherwise the file extension could be checked (like in the demo app).

A preliminary solution could be having a pre-preparation sniff step that will read the first few byes to define the content type (e.g: HLS playlists start with #EXTM3U, DASH manifests should contain an MPD opening tag more or less a the start, etc. falling back on ExtractorMediaSource if all else fails). A proper solution, however, shouldn't use to different connections for this process.

**Comment by regatton** [comment]
> A preliminary solution could be having a pre-preparation sniff step that will read the first few byes to define the content type

Could this function be embedded into ExoPlayer? Something like `GuessMediaSource` that will try to select the most appropriate function to call?

**Comment by AquilesCanta** [comment]
Ideally yes, it should be part of ExoPlayer. My suggestion was aimed towards working around this problem.  We will update this issue once we get to it.

**Comment by sharpaper** [comment]
Is there alternatively another player for Android that can play any Internet audio stream? Perhaps VLC?

**Comment by AquilesCanta** [comment]
I am sure there are, but this probably falls outside the scope of this issue tracker.

**Comment by yvbeek** [comment]
@AquilesCanta @tonihei Are there any plans to include this feature?
It would be great if we no longer have to guess the correct MediaSource type based on the url.

**Comment by tonihei** [comment]
We dropped the plan to add this feature for now under the assumption that the stream type is typically known in the app. 
But we have a pending implementation that solves the issue (and prevents any duplicate data loads): https://github.com/tonihei/ExoPlayer/commit/4a8c860e8c9337c540862df844497c21ec613b57 There was some discussion around the details of this implementation and whether it's the right API approach, but you should be able to use it if you want.

**Comment by timusus** [comment]
I've come up against the same problem - I have an Android client that interacts with various Media Servers (Jellyfin, Emby, Plex).

When requesting audio from these servers, they may respond with HLS streams (without the .m3u8 suffix on the path), or they may respond with binary MPEG content. So the stream type is _not_ known ahead of time.

This is a problem for both ExoPlayer and Chromecast. If the response is HLS and the mimetype is not specified as `x-mpegURL`, the HLS stream fails to play.

I didn't quite follow how to workaround the problem with the link above.. Any advice would be appreciated.

**Comment by tonihei** [comment]
> I didn't quite follow how to workaround the problem with the link above.. Any advice would be appreciated.

The implementation in the link above needs to be patched into the ExoPlayer code to work. Note that the diff is against the code state of July 2020, so you probably need to resolve some merge conflicts if you want to base it on the latest code.

**Comment by timusus** [comment]
@tonihei yes, I understand that. Whart I don't follow is how the diff solves this problem. If you have time, a brief explanation of what's going on there would be super helpful. Edit: I see, the crux of it is the 'PeekDataSource'. I'll have another run at this, thanks.

**Comment by icbaker** [comment]
The other key part of that commit is the `DefaultMediaSource` implementation (which GitHub hides in the diff by default because it's too large). Inside `PeekLoadable` this iterates through `candidateMediaSources` calling `canPrepareWithStream`, which peeks into the stream (as you say, with `PeekDataSource`) to determine if it looks like the right sort of byte sequence. Once it finds a compatible `MediaSource` it delegates to that for subsequent operations.

**Comment by timusus** [comment]
Just wondering whether there have been any updates in this area, or simpler other workarounds. I'm still relying on a patch from back in 2020. 

Edit: Sorry I forgot this repo is deprecated.

**Comment by icbaker** [comment]
Closing this as a duplicate of https://github.com/androidx/media/issues/2898 to move the issue to the media3 issue tracker.
